src/test/fuzz/util.h

```diff
@@ -470,4 +470,63 @@ void ReadFromStream(FuzzedDataProvider& fuzzed_data_provider, Stream& stream) no
     }
 }
 
+class FuzzedSocket : public Socket
+{
+    FuzzedDataProvider& fuzzed_data_provider;
+
+public:
+    explicit FuzzedSocket(FuzzedDataProvider& fuzzed_data_provider_) : fuzzed_data_provider{fuzzed_data_provider_}
+    {
+    }
+
+    ssize_t send(const char* buf, size_t len, int flags) override
+    {
+        if (fuzzed_data_provider.ConsumeBool()) {
+            return len;
+        }
+        return fuzzed_data_provider.ConsumeIntegralInRange<ssize_t>(-1, len);
+    }
+
+    ssize_t recv(char* buf, size_t len, int flags) override
+    {
+        if (buf == nullptr || len == 0 || fuzzed_data_provider.ConsumeBool()) {
+            return fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+        }
+        const std::vector<uint8_t> random_bytes = fuzzed_data_provider.ConsumeBytes<uint8_t>(len);
+        if (random_bytes.empty()) {
+            return fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+        }
+        std::memcpy(buf, random_bytes.data(), random_bytes.size());
+        if (fuzzed_data_provider.ConsumeBool()) {
+            if (len > random_bytes.size()) {
+                std::memset(buf + random_bytes.size(), 0, len - random_bytes.size());
+            }
+            return len;
+        }
+        return random_bytes.size();
```

Comment:
Here it is still possible for `random_bytes.size() < len` since that condition is only checked if `ConsumeBool` returns true. Shouldn't the correct number of bytes be returned if possible?

Reply:
We want to make it so that the fuzzer driven `recv` can return _up to_ `len` bytes like the real `recv`, right? :)

Reply:
Ah, yes that makes sense, thanks for clarifying.